src/Raven.Server/Documents/AttachmentsStorageHelper.cs

```diff
@@ -12,58 +12,58 @@ public static class AttachmentsStorageHelper
-                if(cryptoState.Length < (int)Sodium.crypto_generichash_statebytes())
+                if (cryptoState.Length < (int)Sodium.crypto_generichash_statebytes())
                     throw new InvalidOperationException("BUG: shouldn't happen, the size of a generic hash state was too large!");
 
                 InitComputeHash(cryptoState);
 
                 var bufferRead = 0;
                 while (true)
                 {
-                    var count = await requestStream.ReadAsync(buffer.Buffer.Array, buffer.Buffer.Offset + bufferRead, buffer.Buffer.Count - bufferRead, cancellationToken);
+                    var count = await requestStream.ReadAsync(buffer.Memory.Slice(bufferRead), cancellationToken);
                     if (count == 0)
                         break;
 
                     bufferRead += count;
 
-                    if (bufferRead == buffer.Buffer.Count)
+                    if (bufferRead == buffer.Length)
                     {
                         PartialComputeHash(cryptoState, buffer, bufferRead);
-                        await file.WriteAsync(buffer.Buffer.Array, buffer.Buffer.Offset, bufferRead, cancellationToken);
+                        await file.WriteAsync(buffer.Memory, cancellationToken);
                         bufferRead = 0;
                     }
                 }
-                await file.WriteAsync(buffer.Buffer.Array, buffer.Buffer.Offset, bufferRead, cancellationToken);
+                await file.WriteAsync(buffer.Memory.Slice(0, bufferRead), cancellationToken);
                 PartialComputeHash(cryptoState, buffer, bufferRead);
                 var hash = FinalizeGetHash(cryptoState, buffer);
                 return hash;
             }
         }
 
-        private static unsafe void InitComputeHash(JsonOperationContext.ManagedPinnedBuffer cryptoState)
+        private static unsafe void InitComputeHash(JsonOperationContext.MemoryBuffer cryptoState)
         {
             var rc = Sodium.crypto_generichash_init(cryptoState.Pointer, null, UIntPtr.Zero, Sodium.crypto_generichash_bytes());
             if (rc != 0)
                 throw new InvalidOperationException("Unable to hash attachment: " + rc);
         }
 
-        private static unsafe string FinalizeGetHash(JsonOperationContext.ManagedPinnedBuffer cryptoState, JsonOperationContext.ManagedPinnedBuffer buffer)
+        private static unsafe string FinalizeGetHash(JsonOperationContext.MemoryBuffer cryptoState, JsonOperationContext.MemoryBuffer buffer)
         {
             var size = Sodium.crypto_generichash_bytes();
             var rc = Sodium.crypto_generichash_final(cryptoState.Pointer, buffer.Pointer, size);
             if (rc != 0)
                 throw new InvalidOperationException("Unable to hash attachment: " + rc);
 
-            return Convert.ToBase64String(buffer.Buffer.Array, buffer.Buffer.Offset, (int)size);
+            return Convert.ToBase64String(buffer.Memory.Span.Slice((int)size));
```

Comment:
I think you need this:

```suggestion
            return Convert.ToBase64String(buffer.Memory.Span.Slice(0, (int)size));
```

Reply:
Yes, pushed already, but thanks :)